aten/src/ATen/native/TensorConversions.cpp

```diff
@@ -1795,65 +1846,104 @@ Tensor sparse_compressed_to_sparse(const Tensor& self, c10::optional<c10::Layout
   case kSparseCsc:
     return sparse_compressed_to_sparse_csc(self, dense_dim_opt);
   case kSparseBsr:
-    if (blocksize.has_value()) {
-      return sparse_compressed_to_sparse_bsr(self, *blocksize, dense_dim_opt);
-    } else {
-      const auto blocksize_ = at::sparse_csr::getBlockSize(self);
-      TORCH_CHECK(blocksize_.size() == 2, "to_sparse: ", self.layout(), " to ", layout_,
-                  " conversion requires blocksize specified.");
-      return sparse_compressed_to_sparse_bsr(self, blocksize_, dense_dim_opt);
-    }
+    return sparse_compressed_to_sparse_bsr(
+              self,
+              (self.layout() == kSparseBsc) ?
+                  blocksize.value_or(at::sparse_csr::getBlockSize(self)) :
+                  *blocksize,
+              dense_dim_opt);
   case kSparseBsc:
-    if (blocksize.has_value()) {
-      return sparse_compressed_to_sparse_bsc(self, *blocksize, dense_dim_opt);
-    } else {
-      const auto blocksize_ = at::sparse_csr::getBlockSize(self);
-      TORCH_CHECK(blocksize_.size() == 2, "to_sparse: ", self.layout(), " to ", layout_,
-                  " conversion requires blocksize specified.");
-      return sparse_compressed_to_sparse_bsc(self, blocksize_, dense_dim_opt);
-    }
+    return sparse_compressed_to_sparse_bsc(
+               self,
+               (self.layout() == kSparseBsr) ?
+                   blocksize.value_or(at::sparse_csr::getBlockSize(self)) :
+                   *blocksize,
+               dense_dim_opt);
   default:
     break;
   }
-  AT_ERROR("to_sparse: ", self.layout(), " to ", layout_, " conversion not implemented.");
-  return Tensor();
+
+  AT_ERROR("sparse_compressed_to_sparse: ", self.layout(), " to ", layout_to, " conversion not supported");
+  return Tensor{};
 }
 
 Tensor sparse_coo_to_sparse(const Tensor& self, c10::optional<c10::Layout> layout, OptionalIntArrayRef blocksize, c10::optional<int64_t> dense_dim_opt) {
-  Layout layout_ = layout.value_or(kSparse);
-  TORCH_CHECK(!blocksize.has_value() || layout_ == kSparseBsr || layout_ == kSparseBsc,
-              "to_sparse: ", self.layout(), " to ", layout_,
-              " conversion does not use the specified blocksize ", blocksize.value(), ".");
-  if (dense_dim_opt.has_value()) {
-    AT_ERROR("sparse_coo_to_sparse for ", self.layout(), " to ", layout_, " conversion does not support specifying number of dense dimensions");
-  }
-  if (self.layout() == layout_) {
-    return self;
-  }
-  switch (layout_) {
+  auto layout_to = layout.value_or(kSparse);
+  TORCH_INTERNAL_ASSERT(self.layout() != layout_to, "sparse_coo_to_sparse: same input and output layouts are not supported");
```

Comment:
Nit: this assert can be moved under `case kSparse`.

Reply:
Here I'd prefer to keep it at the top, as it is this way for other functions too.

Reply:
That's fine with me.